Assets/Scripts/Events/WeakVoidPublisher.cs

```diff
@@ -0,0 +1,37 @@
+using System;
+using System.Collections.Generic;
+
+namespace Events
+{
+    // Publisher that does not keep subscribers alive if they are only referenced by this publisher
+    // IMPORTANT: See comment for Publish()
+    // TODO: Currently there is a mild memory leak since the WeakReference objects themselves never get purged.
+    // TODO: This may also impact performance if the simulation runs for a very long time
+    public class WeakVoidPublisher
+    {
+        // save weak references to actions
+        private HashSet<WeakReference<Action>> _subscribers { get; } = new HashSet<WeakReference<Action>>();
+
+        public void Subscribe(Action action) => _subscribers.Add(new WeakReference<Action>(action));
+
+        // TODO: check if this actually works (not sure whether the new object will match the existing one from the set)
+        public bool Unsubscribe(Action action) => _subscribers.Remove(new WeakReference<Action>(action));
+        
+        // IMPORTANT: Don't subscribe Actions that may unreference other subscribed actions.
+        // Otherwise its weak reference and therefore the set will be mutated during iteration.
+        public void Publish()
+        {
+            try
+            {
+                foreach (var weakReference in _subscribers)
+                    if (weakReference.TryGetTarget(out var action))
+                        action.Invoke();
+            }
+            catch (InvalidOperationException e)
+            {
+                throw new Exception("Subscribers to WeakVoidPublisher must never modify" +
+                                    " other subscribers to the same WeakVoidPublisher", e);
+            }
+        }
+    }
```

Comment:
Problem: Autos bleiben sofort für immer stehen. Schwer im Debugger reproduzierbar, da sich die Autos beim Debuggen erst etwas bewegen, bevor sie auch für immer stehen bleiben.

Reply:
Welchen Vorteil hat denn das WeakRefefencing gegenüber der vorherigen Methode?

Reply:
Zeile 6:
```cs
// Publisher that does not keep subscribers alive if they are only referenced by this publisher
```
Die Autos werden sonst von den Referenzen im Publisher für immer am Leben erhalten, und der Publisher ruft dann z.B. Move() bei allen Autos auf, die eigentlich schon despawned sind und deren GameObject schon gelöscht (null) ist.

Die schwachen Referenzen halten die Auto Objekte nicht am Leben und wenn ein Auto zerstört wird, wird das Target von der WeakReference einfach zu null.

Reply:
Wenn die Autos gelöscht werden, wird ja ihre `Dispose()` Funktion aufgerufen, die jedes GameObjectData hat. Dann kann man sich doch in dieser Funktion von allen Events unsubscriben.

Reply:
Das wäre die alternative Lösung. Ist halt sehr hardgecoded dann.

Ich glaube man kann das ganze System etwas angenehmer lösen:
UpdatePublisher speichert irgendnen `OrderedDictionary<Type, OrderedDictionary<object, events>>`.

Beim Subscriben muss das eigene Objekt und die Events eingegeben werden.
Zusätzlich kann man spezifizieren, dass die eigenen Events VOR oder NACH den Events von einem beliebigen anderen Datentyp aufgerufen werden (deshalb OrderedDictionary).

In Dispose() werden einfach alle Events vom eigenen Objekt automatisch gelöscht.

Werde das morgen implementieren, dann wirds vielleicht klarer.